### DISCUSSION TITLE
Enhancement: Add more tools for assistants and agents.

### DISCUSSION THREAD

**Discussion by Passerby1011** [danny-avila/LibreChat#4524] [Feature Requests & Suggestions]
### What features would you like to see added?

As the title suggests, add more tools for assistants and agents, such as the very important web browsing and code interpreter.

### More details

Many of the tools in gptplugins are currently non-functional. Are there plans to update them or remove gptplugins entirely?
It is similar to agents in functionality, but it is more flexible to use than agents.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by jmaddington** [comment]
I have a web browser tool at https://github.com/jmaddington/LibreChat/tree/new/feature/curl

Danny released a code interpreter since you wrote this comment. 

My branch above isn't up to date with the late code but I could update it if it were useful

**Reply by Passerby1011** [reply]
I think this is very helpful. Currently, I rely on third-party tools for web browsing, which is not as convenient as having it built-in. Moreover, LibreChat needs such a tool, and you can submit a PR.

**Reply by jmaddington** [reply]
It's on my todo list. I have something like 3 to create. Flux, WebNavigator and Open Weather (in progress).

**Reply by Passerby1011** [reply]
> It's on my todo list. I have something like 3 to create. Flux, WebNavigator and Open Weather (in progress).

Thank you very much for your work
At the same time, I would like to make some small comments of my own,
For models like flux that generate images from text, I don't think we should adapt them individually; my current approach is to write a separate action for each manufacturer (using AI generation). Much of this is repetitive. The way that needs to be changed is request addresses, request parameters, parameter specifications. I think you could make this an open plugin if you could. The user could configure the model themselves, if librechat supports it.

**Reply by jmaddington** [reply]
You mean a single plugin that supports all image generation AI models at once?

**Reply by Passerby1011** [reply]
Yes, there are a growing number of models currently supporting image generation. It would be too slow and too much to adapt one by one. The main parameters of these image generation models are prompt, model, and size.
And the nature of plugins in librechat is the same as actions. The plug-in is just a little more convenient，
It's just difficult to put all the image generation models in a single plugin and requires a lot of customization options

**Reply by kneelesh48** [reply]
There are multiple api providers for image generation, such as openai, google (soon), replicate, together, fal.ai etc. We can have a similar system where you choose the provider (endpoint), model and then things such as image dimensions can be provided in settings. When we ask an LLM to generate an image, it provides the prompt and number of images and we can have a setting to have the LLM rewrite the setting.

**Reply by Passerby1011** [reply]
In fact, many Chinese manufacturers are also working on AI large model image generation, such as Doubao, Qwen, Zhupu, SiliconFlow, and Tencent. This has been quite a headache for me, but fortunately, their request parameters are mostly similar, typically including prompt, model, size, and seed (optional).
